File: apps/python/blueprints/exceptions.py
import azure.functions as func
import json
import logging

bp = func.Blueprint()
logger = logging.getLogger(__name__)


@bp.route(route="exceptions/test-error", methods=["GET"])
def test_error(req: func.HttpRequest) -> func.HttpResponse:
    """Raise and catch a ValueError to demonstrate exception handling."""
    logger.info("Exception test endpoint requested")

    try:
        raise ValueError("Simulated error for testing")
    except ValueError as exc:
        logger.exception("Caught simulated error: %s", exc)

        body = json.dumps({
            "error": "Handled exception",
            "type": type(exc).__name__,
            "message": str(exc),
        })
        return func.HttpResponse(body, mimetype="application/json", status_code=200)

File: apps/nodejs/src/functions/exceptions.js
const { app } = require('@azure/functions');

/**
 * Raise and catch an error to demonstrate exception handling.
 */
app.http('testError', {
    methods: ['GET'],
    route: 'exceptions/test-error',
    handler: async (request, context) => {
        context.log('Exception test endpoint requested');

        try {
            throw new Error('Simulated error for testing');
        } catch (err) {
            context.error(`Caught simulated error: ${err.message}`);

            return {
                status: 200,
                jsonBody: {
                    error: 'Handled exception',
                    type: err.constructor.name,
                    message: err.message,
                },
            };
        }
    },
});

/**
 * Raise an unhandled exception to generate real 500 errors in Application Insights.
 */
app.http('unhandledError', {
    methods: ['GET'],
    route: 'exceptions/unhandled',
    handler: async (request, context) => {
        context.log('Unhandled exception endpoint requested');
        throw new Error('Deliberate unhandled error for KQL data collection');
    },
});
